### DISCUSSION TITLE
How to write tests?

### DISCUSSION THREAD

**Discussion by punksta** [invertase/react-native-firebase#5771] [Q&A] [answered]
Thank you for a great library! 

I have module for firebase firestore integration which I need to test. I am using batch queries, security rules, transactions, etc... 
I want to be confident that module follows product requirements. 

How can I test it? Preferably in e2e with real firebase environment. I have development project in firebase which I can use for testing.

I am using detox and jest in my project.

**Comment by mikehardy** [comment, marked as answer]
Hi there! This is not firebase specific I think. With the details provided you could ask the same question about an app that uses CouchDB or similar. Design your screens so they have testIDs etc in the elements, and drive your app with Detox, asserting where necessary. It should be as simple (or as complicated) as that.

**Reply by punksta** [reply]
Thanks for your answer.

**Reply by punksta** [reply]
I wanted to have more control in the tests and test specific functions of a isolated module. I didn't mean UI tests.
I want to run them as close to real environment as possible, that's what I mean by e2e.

I can see that you are using `jet` to run your tests which run on emulator. You also recommend to not use `jet`elsewhere. That's why I asked this question.

**Reply by mikehardy** [reply]
That's important context :-). Yeah, jet currently uses a low-level react-native interface (the "Javascript Debugging" interface) that is going away in the future in react-native. So I'm migrating to a different style on a jet-next branch over there and I'm still setting on an API for test authoring while I bring up the new harness style.

For you now, in practical terms, make e2e-specific screens that contain simple buttons for the things you want to test I think, and drive those with Detox, or have the screens themselves just run the tests on load with a text that says "in progress" or "success" or "failure" and have detox wait for it to resolve to success or failure perhaps.

**Reply by punksta** [reply]
@mikehardy thank you for your idea and good luck with redesign of the tests 👍
